File: src/main/java/io/github/oleksivio/telegram/bot/core/controller/annotations/parser/producer/ArgCheckerProducer.java
package io.github.oleksivio.telegram.bot.core.controller.annotations.parser.producer;

import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.filter.FilterParser;
import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.finder.Finder;
import io.github.oleksivio.telegram.bot.core.controller.handler.check.Validator;
import io.github.oleksivio.telegram.bot.core.model.ITelegram;
import org.springframework.stereotype.Controller;

import java.lang.annotation.Annotation;
import java.lang.reflect.Method;
import java.util.Collection;
import java.util.Optional;

@Controller
public class ArgCheckerProducer {
    private final Collection<FilterParser> filterParsers;
    private final Finder finder;

    public ArgCheckerProducer(Collection<FilterParser> filterParsers, Finder finder) {
        this.filterParsers = filterParsers;
        this.finder = finder;
    }

    public <A extends Annotation, T extends ITelegram> Optional<Validator<T>> create(Method method) {

        for (FilterParser<A, T> filterParser : filterParsers) {
            if (method.isAnnotationPresent(filterParser.getAnnotationClass())) {
                A annotation = method.getAnnotation(filterParser.getAnnotationClass());
                Validator<T> validator = filterParser.createChecker(annotation, finder);
                return Optional.of(validator);
            }

        }

        return Optional.empty();
    }

}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/core/controller/annotations/parser/producer/ArgCheckerProducer.kt
package io.github.oleksivio.telegram.bot.core.controller.annotations.parser.producer

import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.filter.FilterParser
import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.finder.Finder
import io.github.oleksivio.telegram.bot.core.controller.handler.check.Validator
import io.github.oleksivio.telegram.bot.core.model.ITelegram
import org.springframework.stereotype.Controller
import kotlin.reflect.KFunction

/**
 * Created by oleksivio on 15.03.19 at 00:21
 * Project: telegram-bot-api
 */
@Controller
class ArgCheckerProducer(private val filterParsers: Collection<FilterParser<*, *>>, private val finder: Finder) {

    fun <A : Annotation, T : ITelegram> create(func: KFunction<*>): Validator<T> {

        for (filterParser in filterParsers) {
            val annotationAbstract = func.annotations
                    .firstOrNull { it.annotationClass == filterParser.parserAnnotationClass }
            if (annotationAbstract != null) {
                val annotation = annotationAbstract as A
                val filter = filterParser as FilterParser<A, T>
                return filter.createChecker(annotation, finder)
            }

        }

        return { true }
    }

}
